### PR TITLE
[VGE-825] Live Reloading Support for zcli apps:server

### PR DESCRIPTION
**Pull request by etanxing**
## Description

When `manifest.json` is changed the server will reflect changes in it without stopping & restarting manually.

## Detail

VEG-825

## Checklist

- [x] :guardsman: includes new unit and functional tests

### CODE REVIEW

File: packages/zcli-apps/src/commands/apps/server.ts [hunk 1]
```diff
@@ -53,14 +54,39 @@ export default class Server extends Command {
...
+
+    // Keep references of watchers for unwatching later
+    const watchers = appPaths.map(appPath =>
+      fs.watch(appPath, async (eventType, filename) => {
+        if (filename.toLowerCase() === 'manifest.json') {
+          // Regenerate app.json
+          appJSON = await buildAppJSON(appPaths, port)
+          // Reset middlewares for app assets
+          setAppAssetsMiddleware()
+          this.log('manifest.json changed and please refresh.')
```
Comment (token-cjg) [inline packages/zcli-apps/src/commands/apps/server.ts:71]: Maybe reword this slightly to `The manifest of the served app has changed, please refresh in your preferred host (eg Support or Sell) if you wish to pick up the changes.`

File: packages/zcli-apps/tests/functional/server.test.ts [hunk 2]
```diff
@@ -84,4 +85,34 @@ describe('apps server', function () {
...
+    let server: http.Server, appHost: string, appJSON: AppJSONPayload
+    before(async () => {
+      server = await ServerCommand.run([singleProductApp])
+      appHost = 'http://localhost:4567'
+    })
+
+    after(() => server.close())
+
+    test
+      .it('should reflect changes in mainfest.json', async () => {
```
Comment (token-cjg) [inline packages/zcli-apps/tests/functional/server.test.ts:100]: nit, `mainfest` -> `manifest`

File: packages/zcli-apps/tests/functional/server.test.ts [hunk 3]
```diff
@@ -84,4 +85,34 @@ describe('apps server', function () {
...
+        // Read mainfest.json
+        const mainfestPath = path.join(singleProductApp, 'manifest.json')
```
Comment (token-cjg) [inline packages/zcli-apps/tests/functional/server.test.ts:102]: Another couple of minor typos here of the same ilk

File: packages/zcli-apps/tests/functional/server.test.ts [hunk 4]
```diff
@@ -84,4 +85,34 @@ describe('apps server', function () {
...
+        const manifest: Manifest = JSON.parse(fs.readFileSync(mainfestPath, 'utf8'))
+        const appName = manifest.name
+        manifest.name = `${appName} modified`
+        // Modifed mainfest.json
+        fs.writeFileSync(mainfestPath, JSON.stringify(manifest))
+        await wait(200)
```
Comment (token-cjg) [inline packages/zcli-apps/tests/functional/server.test.ts:108]: Would it be possible to avoid an explicit wait here? Maybe consider an await as prefix to `fs.writeFileSync` on the line above instead?

**Review by token-cjg — COMMENTED** [review]
Looks good, maybe we could add a testcase for the situation where we might have multiple watchers, rather than just one as per your current unit test? I vaguely recall that zcli can serve more than one app at a time --- not essential but just a thought.

File: packages/zcli-apps/src/commands/apps/server.ts (hunk 1 above)
Comment (etanxing) [reply packages/zcli-apps/src/commands/apps/server.ts:71]: Yup, I just feel that it's a little bit long. I've asked @zach-anthony to see if he has any preferred wording.

**Comment by etanxing** [comment]
> Looks good, maybe we could add a testcase for the situation where we might have multiple watchers, rather than just one as per your current unit test? I vaguely recall that zcli can serve more than one app at a time --- not essential but just a thought.

sure, will add

File: packages/zcli-apps/src/commands/apps/server.ts (hunk 1 above)
Comment (etanxing) [reply packages/zcli-apps/src/commands/apps/server.ts:71]: removed the log as per @zach-anthony suggested.

**Review by cl8au — APPROVED** [review]

**Comment by etanxing** [comment]
> Looks good, maybe we could add a testcase for the situation where we might have multiple watchers, rather than just one as per your current unit test? I vaguely recall that zcli can serve more than one app at a time --- not essential but just a thought.

multiple watchers tests added

File: packages/zcli-apps/tests/functional/server.test.ts (hunk 4 above)
Comment (etanxing) [reply packages/zcli-apps/tests/functional/server.test.ts]: `fs.writeFileSync` returns nothing (void), we can't really `await` as `await` need a promise. I reduced the wait time to `10 ms`.

Comment (token-cjg) [reply packages/zcli-apps/tests/functional/server.test.ts]: I think the approach here would be to use the promises module in fs, [see this stackoverflow answer](https://stackoverflow.com/a/52094177) for instance.  I remember now, that is why we have those weird imports of the promise submodule as fs in other parts of zcli - it is precisely because of this issue.

So you'd do something like this

```
import { promises as fs } from 'fs';

(async () => {
    [...preceding code...]
    await fs.writeFile(mainfestPath, JSON.stringify(manifest), 'utf8');
    [...following code...]
})();
```

Comment (etanxing) [reply packages/zcli-apps/tests/functional/server.test.ts]: Yes, I did `import { promises as fs } from 'fs';` with chokidar but `await wait()` is still required. Because we need to regenerate `app.json` and reset the middlewares  after `manifest.json` changed. Without waiting we get the old `app.json` as resetting the middlewares is done after we fetched `app.json` (based on logs I added).

File: packages/zcli-apps/src/commands/apps/server.ts [hunk 5]
```diff
@@ -53,14 +54,38 @@ export default class Server extends Command {
...
-    })
+    const setAppAssetsMiddleware = () => {
+      appJSON.installations.forEach((installation: Installation, index: number) => {
+        app.use(`/${installation.app_id}/assets`, express.static(`${appPaths[index]}/assets`))
+      })
+    }
+
+    // Keep references of watchers for unwatching later
+    const watchers = appPaths.map(appPath =>
+      fs.watch(appPath, async (eventType, filename) => {
```
Comment (sarmadsangi) [inline packages/zcli-apps/src/commands/apps/server.ts:65]: Overall looks great. Few questions,

1. Any specific reason we are not using cross-platform lib like Chokidar here? I think it solves many inconsistencies between various platforms.

2. Also, should we consider ignoring `node_modules` directory? I think it might be a good idea to ignore that to avoid any potential perf issues (recursive nature of node_modules, and can sometimes be millions of files)

File: packages/zcli-apps/src/commands/apps/server.ts [hunk 6]
```diff
@@ -53,14 +54,38 @@ export default class Server extends Command {
...
       this.log(`Add ${chalk.bold('?zcli_apps=true')} to the end of your Zendesk URL to load these apps on your Zendesk account.\n`)
       tailLogs && this.log(chalk.bold('Tailing logs'))
     })
+
+    return {
+      close: () => {
+        // Stop watching file changes before terminating the server
+        watchers.forEach(watcher => watcher.close())
+        server.close()
+      }
```
Comment (sarmadsangi) [inline packages/zcli-apps/src/commands/apps/server.ts:88]: 👏 nice clean up

File: packages/zcli-apps/src/commands/apps/server.ts (hunk 5 above)
Comment (token-cjg) [reply packages/zcli-apps/src/commands/apps/server.ts:65]: Chokidar was used at first I believe but tests failed on Linux and Windows.

Comment (etanxing) [reply packages/zcli-apps/src/commands/apps/server.ts:65]: 1. Yup, we tried Chokidar but [tests failed on Linux and Windows](https://github.com/zendesk/zcli/actions/runs/1460230663). 
2. What do you mean by ignoring `node_modules` directory? It's in `.gitignore`.
Thanks

File: packages/zcli-apps/tests/functional/server.test.ts (hunk 4 above)
Comment (token-cjg) [reply packages/zcli-apps/tests/functional/server.test.ts]: I'm not entirely convinced, but I won't withhold a +1 on this PR if you believe that this is a hard blocker to removing the use of `wait()` here.  `wait` is definitely a smell though which it would be good to expurgate from the code in a follow up task.

**Review by token-cjg — APPROVED** [review]

File: packages/zcli-apps/tests/functional/server.test.ts (hunk 4 above)
Comment (token-cjg) [reply packages/zcli-apps/tests/functional/server.test.ts]: I've created a followup task and childed it to the current one in the sprint.

File: packages/zcli-apps/src/commands/apps/server.ts (hunk 5 above)
Comment (sarmadsangi) [reply packages/zcli-apps/src/commands/apps/server.ts:65]: @etanxing not our `node_modules`, 3rd party apps `node_modules`.

Do we know why chokidar failed? could be a real test failure or pointing to some flaky tests.

Comment (etanxing) [reply packages/zcli-apps/src/commands/apps/server.ts:65]: The chokidar watcher wasn’t triggered when `manifest.json` was changed on Linux and Windows. On MacOS, the tests succeed so there is a cross platform issue with chokidar.
